packages/catalog-model/src/location/parser.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LocationSpec, LOCATION_TYPES, LocationType } from './';
+import path from 'path';
+import os from 'os';
+import fs from 'fs';
+import parseGitUrl from 'git-url-parse';
+import { Clone, Repository } from 'nodegit';
+
+
+/**
+ * A policy for validation or mutation to be applied to entities as they are
+ * entering the system.
+ */
+type LocationImpl = {
+  /**
+   * Parses and validates a location string.
+   *
+   * @returns A LocationSpec type object
+   * @throws An error if the location is not valid
+   */
+  parse(): LocationSpec;
```

Comment:
Thinking this would be modelled as `Location.parse()` instead, and let `Location` extend `LocationSpec` in read-only form. The constructor would just take a LocationSpec instead, possibly validating it.